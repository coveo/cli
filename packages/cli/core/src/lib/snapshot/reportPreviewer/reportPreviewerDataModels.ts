import {ResourceSnapshotsReportOperationModel} from '@coveo/platform-client';

export type ReportViewerOperationName =
  keyof ResourceSnapshotsReportOperationModel;

export interface ReportViewerResourceReportModel {
  name: string;
  operations: ResourceSnapshotsReportOperationModel;
}

export enum SnapshotReportStatus {
  SUCCESS = 'SUCCESS',
  NO_CHANGES = 'NO_CHANGES',
  MISSING_VAULT_ENTRIES = 'MISSING_VAULT_ENTRIES',
  ERROR = 'ERROR',
}
