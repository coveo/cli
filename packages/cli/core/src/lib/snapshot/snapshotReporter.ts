import {
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportOperationModel,
  ResourceSnapshotsReportOperationResult,
  ResourceSnapshotType,
} from '@coveo/platform-client';
import {
  ReportViewerOperationName,
  ReportViewerResourceReportModel,
  SnapshotReportStatus,
} from './reportPreviewer/reportPreviewerDataModels';

type ResourceEntries = [string, ResourceSnapshotsReportOperationModel];

type NoopHandler = (this: void) => void;
type SnapshotReporterHandler = (this: SnapshotReporter) => void | Promise<void>;

type SnapshotReporterHandlers = {
  [K in SnapshotReportStatus]: SnapshotReporterHandler | NoopHandler;
};

const MissingVaultEntryResultCode = 'INVALID_PLACEHOLDER';
//#endregion

export type VaultEntryAttributes = {
  vaultEntryId: string;
  resourceName: string;
  resourceType: ResourceSnapshotType;
};

export class SnapshotReporter {
  private missingVaultEntriesSet: Set<VaultEntryAttributes> = new Set();

  private operationResultsParseFuseUsed: boolean = false;

  private consumeOperationResultsParseFuse() {
    if (this.operationResultsParseFuseUsed) {
      return false;
    } else {
      this.operationResultsParseFuseUsed = true;
      return true;
    }
  }

  public get missingVaultEntries() {
    return this.missingVaultEntriesSet.values();
  }
  public resourceInErrorCount: number = 0;
  public resourceInError: Map<Partial<ResourceSnapshotType>, Set<string>> =
    new Map();
  public constructor(public readonly report: ResourceSnapshotsReportModel) {}

  public static convertResourceEntriesToResourceSnapshotsReportOperationModel([
    name,
    operations,
  ]: ResourceEntries) {
    return {name, operations};
  }

  public static resourceHasAtLeastOneOperationFilter(
    operationsToDisplay: ReportViewerOperationName[]
  ) {
    return ([_, operations]: ResourceEntries) =>
      operationsToDisplay.reduce(
        (previous, current) => previous + operations[current],
        0
      ) > 0;
  }

  public getAffectedResourceTypes(): ResourceSnapshotType[] {
    const isResourceTypeReportAffected = (
      report: ResourceSnapshotsReportOperationModel
    ) => Object.values(report).some((operation) => operation > 0);
    const changedResourcesTypes: ResourceSnapshotType[] = [];
    for (const resourceType of Object.keys(this.report.resourceOperations)) {
      if (
        isResourceTypeReportAffected(
          this.report.resourceOperations[resourceType]
        )
      ) {
        changedResourcesTypes.push(resourceType as ResourceSnapshotType);
      }
    }
    return changedResourcesTypes;
  }

  public hasChangedResources() {
    const totalUnchanges =
      this.getOperationTypeTotalCount('resourcesUnchanged');
    return totalUnchanges !== this.report.resourcesProcessed;
  }

  public getOperationTypeTotalCount(
    type: keyof ResourceSnapshotsReportOperationModel
  ) {
    const count = Object.values(this.report.resourceOperations).reduce(
      (count: number, current: ResourceSnapshotsReportOperationModel) =>
        count + current[type],
      0
    );

    return count;
  }

  private reportHandlers: SnapshotReporterHandlers = {
    [SnapshotReportStatus.SUCCESS]: () => {},
    [SnapshotReportStatus.NO_CHANGES]: () => {},
    [SnapshotReportStatus.MISSING_VAULT_ENTRIES]: () => {},
    [SnapshotReportStatus.ERROR]: () => {},
  };

  public setReportHandler<T extends SnapshotReportStatus>(
    status: T,
    handler: SnapshotReporterHandlers[T]
  ): SnapshotReporter {
    this.reportHandlers[status] = handler;
    return this;
  }

  public async handleReport(): Promise<void> {
    const reportStatuses = this.getReportStatuses();
    await this.executeHandlers(reportStatuses.fixables);
    if (reportStatuses.errors.length > 0) {
      await this.executeHandlers(reportStatuses.errors);
      return;
    }
    if (reportStatuses.successes.length > 0) {
      await this.executeHandlers(reportStatuses.successes);
    }
  }

  private async executeHandlers(reportStatuses: SnapshotReportStatus[]) {
    for (const status of reportStatuses) {
      await this.callAndReset(status);
    }
  }

  private async callAndReset(handler: SnapshotReportStatus) {
    await this.reportHandlers[handler].apply(this);
    this.reportHandlers[handler] = () => {};
  }

  private getReportStatuses() {
    const statuses: Record<
      'fixables' | 'successes' | 'errors',
      SnapshotReportStatus[]
    > = {
      fixables: [],
      successes: [],
      errors: [],
    };

    this.parseResourcesOperationsResults();
    if (this.missingVaultEntriesSet.size > 0) {
      statuses.fixables.push(SnapshotReportStatus.MISSING_VAULT_ENTRIES);
    }

    if (!this.hasChangedResources()) {
      statuses.successes.push(SnapshotReportStatus.NO_CHANGES);
    }

    if (this.resourceInErrorCount > 0) {
      statuses.errors.push(SnapshotReportStatus.ERROR);
    } else {
      statuses.successes.push(SnapshotReportStatus.SUCCESS);
    }

    return statuses;
  }

  private parseResourcesOperationsResults(): void {
    if (!this.consumeOperationResultsParseFuse()) {
      return;
    }
    for (const [resourceType, resource] of Object.entries(
      this.report.resourceOperationResults
    )) {
      for (const [resourceName, errors] of Object.entries(resource)) {
        for (const err of errors) {
          this.parseResourceOperationResult(err, resourceName, resourceType);
        }
      }
    }
  }

  private parseResourceOperationResult(
    err: ResourceSnapshotsReportOperationResult,
    resourceName: string,
    resourceType: string
  ) {
    const missingEntry = SnapshotReporter.tryGetMissingVaultEntryName(err);
    if (missingEntry) {
      this.missingVaultEntriesSet.add({
        vaultEntryId: missingEntry,
        resourceName: resourceName,
        resourceType: resourceType as ResourceSnapshotType,
      });
    } else {
      // TODO: Fix PlatformClient to reflect proper typing.
      this.addResourceInError(resourceType as ResourceSnapshotType, err);
    }
  }

  private addResourceInError(
    resourceType: ResourceSnapshotType,
    err: ResourceSnapshotsReportOperationResult
  ) {
    this.resourceInErrorCount++;
    let errorSet = this.resourceInError.get(resourceType);
    if (!errorSet) {
      errorSet = new Set();
      this.resourceInError.set(resourceType, errorSet);
    }
    errorSet.add(`${err.resultCode}: ${err.message}`);
  }

  private static missingVaultEntryMatcher =
    /^The vault entry referenced by \{\{ VAULT\.(?<entryName>.*) \}\} could not be found in the vault\.$/;

  private static tryGetMissingVaultEntryName(
    err: ResourceSnapshotsReportOperationResult
  ): string | undefined {
    let message: string = '';
    if (typeof err !== 'string') {
      if (err.resultCode === MissingVaultEntryResultCode) {
        message = err.message;
      } else {
        return undefined;
      }
    } else {
      message = err;
    }
    const match = message.match(SnapshotReporter.missingVaultEntryMatcher);
    return match?.groups?.['entryName'];
  }

  public getChangedResources(
    operationsToDisplay: ReportViewerOperationName[]
  ): ReportViewerResourceReportModel[] {
    return Object.entries(this.report.resourceOperations)
      .filter(
        SnapshotReporter.resourceHasAtLeastOneOperationFilter(
          operationsToDisplay
        )
      )
      .map(
        SnapshotReporter.convertResourceEntriesToResourceSnapshotsReportOperationModel
      );
  }

  public get type() {
    const type = this.prettify(this.report.type);
    return type;
  }

  public get status() {
    const status = this.prettify(this.report.status);
    return status;
  }

  public get resultCode() {
    const status = this.prettify(this.report.resultCode);
    return status;
  }

  private prettify(str: string): string {
    const capitalized = (word: string) =>
      word.charAt(0) + word.slice(1).toLowerCase();

    const spaced = str.replace(/_/g, ' ');
    return spaced.split(' ').map(capitalized).join(' ');
  }
}
