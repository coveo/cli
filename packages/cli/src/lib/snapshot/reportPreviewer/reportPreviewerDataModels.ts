import {ResourceSnapshotsReportOperationModel} from '@coveord/platform-client';

export type ReportViewerOperationName =
  keyof ResourceSnapshotsReportOperationModel;

export interface ReportViewerResourceReportModel {
  name: string;
  operations: ResourceSnapshotsReportOperationModel;
}
