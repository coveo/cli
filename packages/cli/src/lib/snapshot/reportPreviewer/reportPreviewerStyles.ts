import chalk from 'chalk';

export const ReportViewerStyles = {
  green: (txt: string) => chalk.green(txt),
  yellow: (txt: string) => chalk.yellow(txt),
  red: (txt: string) => chalk.red(txt),
  gray: (txt: string) => chalk.gray(txt),
  header: (txt: string) => chalk.bold.hex('#1CEBCF')(txt),
  error: (txt: string) => chalk.bgHex('#F64D64').hex('#272C3A')(txt),
};
