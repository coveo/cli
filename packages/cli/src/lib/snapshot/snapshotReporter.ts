import {
  ResourceSnapshotsReportModel,
  ResourceSnapshotsReportOperationModel,
  ResourceSnapshotsReportResultCode,
  ResourceSnapshotsReportStatus,
} from '@coveord/platform-client';
import {
  ReportViewerOperationName,
  ReportViewerResourceReportModel,
} from './reportPreviewer/reportPreviewerDataModels';

type ResourceEntries = [string, ResourceSnapshotsReportOperationModel];

export class SnapshotReporter {
  public constructor(public readonly report: ResourceSnapshotsReportModel) {}

  public static convertResourceEntriesToResourceSnapshotsReportOperationModel([
    name,
    operations,
  ]: ResourceEntries) {
    return {name, operations};
  }

  public static resourceHasAtLeastOneOperationFilter(
    operationsToDisplay: ReportViewerOperationName[]
  ) {
    return ([_, operations]: ResourceEntries) =>
      operationsToDisplay.reduce(
        (previous, current) => previous + operations[current],
        0
      ) > 0;
  }

  public hasChangedResources() {
    const totalUnchanges =
      this.getOperationTypeTotalCount('resourcesUnchanged');
    return totalUnchanges !== this.report.resourcesProcessed;
  }

  public getOperationTypeTotalCount(
    type: keyof ResourceSnapshotsReportOperationModel
  ) {
    const count = Object.values(this.report.resourceOperations).reduce(
      (count: number, current: ResourceSnapshotsReportOperationModel) =>
        count + current[type],
      0
    );

    return count;
  }

  public isSuccessReport(): boolean {
    const {status, resultCode} = this.report;
    return (
      status === ResourceSnapshotsReportStatus.Completed &&
      resultCode === ResourceSnapshotsReportResultCode.Success
    );
  }

  public getChangedResources(
    operationsToDisplay: ReportViewerOperationName[]
  ): ReportViewerResourceReportModel[] {
    return Object.entries(this.report.resourceOperations)
      .filter(
        SnapshotReporter.resourceHasAtLeastOneOperationFilter(
          operationsToDisplay
        )
      )
      .map(
        SnapshotReporter.convertResourceEntriesToResourceSnapshotsReportOperationModel
      );
  }

  public get type() {
    const type = this.prettify(this.report.type);
    return type;
  }

  public get status() {
    const status = this.prettify(this.report.status);
    return status;
  }

  public get resultCode() {
    const status = this.prettify(this.report.resultCode);
    return status;
  }

  private prettify(str: string): string {
    const capitalized = (word: string) =>
      word.charAt(0) + word.slice(1).toLowerCase();

    const spaced = str.replace(/_/g, ' ');
    return spaced.split(' ').map(capitalized).join(' ');
  }
}
