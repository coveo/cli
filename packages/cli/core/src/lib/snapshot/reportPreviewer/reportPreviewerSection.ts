import {ReportViewerStyles} from './reportPreviewerStyles';
import {
  ReportViewerOperationName,
  ReportViewerResourceReportModel,
} from './reportPreviewerDataModels';
import {ResourceSnapshotsReportOperationModel} from '@coveo/platform-client';

class ReportViewerOperationLogFactory {
  public constructor(
    private operator: string,
    private color: keyof typeof ReportViewerStyles,
    private templateString: (count: number) => string
  ) {}

  public getString(count: number, indentation: number) {
    return `${ReportViewerStyles[this.color](
      this.operator.padEnd(indentation)
    )}${ReportViewerStyles[this.color](this.templateString(count))}\n`;
  }
}

export class ReportViewerSection {
  public constructor(
    private resource: ReportViewerResourceReportModel,
    private operationsToDisplay: ReportViewerOperationName[]
  ) {}

  public static operationToLogFactoryDictionnary: {
    [key in ReportViewerOperationName]: ReportViewerOperationLogFactory;
  } = {
    resourcesCreated: new ReportViewerOperationLogFactory(
      '+',
      'green',
      (count) => `${count} to create`
    ),
    resourcesRecreated: new ReportViewerOperationLogFactory(
      '+-',
      'yellow',
      (count) => `${count} to replace`
    ),
    resourcesUpdated: new ReportViewerOperationLogFactory(
      '~',
      'yellow',
      (count) => `${count} to update`
    ),
    resourcesDeleted: new ReportViewerOperationLogFactory(
      '-',
      'red',
      (count) => `${count} to delete`
    ),
    resourcesInError: new ReportViewerOperationLogFactory(
      '!',
      'error',
      (count) => `${count} in error`
    ),
    resourcesUnchanged: new ReportViewerOperationLogFactory(
      ' ',
      'gray',
      (count) => `${count} unchanged`
    ),
  };

  private shouldPrintOperation(
    operation: keyof ResourceSnapshotsReportOperationModel
  ): boolean {
    const canBeDisplayed = this.operationsToDisplay.includes(operation);
    const isResourceAffectedByOperation =
      this.resource.operations[operation] > 0;

    return canBeDisplayed && isResourceAffectedByOperation;
  }

  public display(indentation: number) {
    return this.operationsToDisplay.reduce((previous, current) => {
      if (!this.shouldPrintOperation(current)) {
        return previous;
      }
      const operationToLog =
        ReportViewerSection.operationToLogFactoryDictionnary[current].getString(
          this.resource.operations[current],
          indentation
        );
      return previous + operationToLog;
    }, '');
  }
}
