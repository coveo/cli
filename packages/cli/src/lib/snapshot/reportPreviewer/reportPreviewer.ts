import {CliUx} from '@oclif/core';
import {red, italic, green, yellow, ChalkFunction} from 'chalk';
import {ReportViewerSection} from './reportPreviewerSection';
import {ReportViewerStyles} from './reportPreviewerStyles';
import {SnapshotReporter} from '../snapshotReporter';
import {
  ReportViewerOperationName,
  ReportViewerResourceReportModel,
  SnapshotReportStatus,
} from './reportPreviewerDataModels';
import dedent from 'ts-dedent';
import {recordable} from '../../utils/record';
import {ResourceSnapshotType} from '@coveord/platform-client';

type Plurable = [singular: string, plural: string];
function pluralizeIfNeeded(plurable: Plurable, unprintedMessages: number) {
  return plurable[unprintedMessages > 1 ? 1 : 0];
}

export class ReportViewer {
  private static errorPlurable: Plurable = ['error', 'errors'];
  private static entryPlurable: Plurable = ['entry', 'entries'];
  private static resourcePlurable: Plurable = ['resource', 'resources'];

  public static defaultOperationsToDisplay: ReportViewerOperationName[] = [
    'resourcesCreated',
    'resourcesDeleted',
    'resourcesInError',
    'resourcesRecreated',
    'resourcesUnchanged',
    'resourcesUpdated',
  ];

  public static maximumNumberOfErrorsToPrint = 5;

  private operationsToDisplay: ReportViewerOperationName[];

  public constructor(
    private readonly reporter: SnapshotReporter,
    operationsToDisplay: ReportViewerOperationName[] = []
  ) {
    this.operationsToDisplay =
      ReportViewer.defaultOperationsToDisplay.concat(operationsToDisplay);
  }

  public async display(): Promise<void> {
    this.printTable();
    await this.reporter
      .setReportHandler(
        SnapshotReportStatus.ERROR,
        this.getReportErrorsHandler()
      )
      .setReportHandler(
        SnapshotReportStatus.NO_CHANGES,
        this.getNoChangesReportHandler()
      )
      .setReportHandler(
        SnapshotReportStatus.MISSING_VAULT_ENTRIES,
        this.getMissingEntriesHandler()
      )
      .handleReport();
  }

  private getMissingEntriesHandler(): (
    this: SnapshotReporter
  ) => void | Promise<void> {
    const missingVaultEntryPrinter = (entries: string[]) =>
      ReportViewer.printAbridgedMessages(
        entries,
        ReportViewer.entryPlurable,
        yellow
      );
    return function (this: SnapshotReporter) {
      const entries = Array.from(this.missingVaultEntries).map(
        (entry) => entry[0]
      );
      CliUx.ux.log(
        yellow(
          `Missing vault ${pluralizeIfNeeded(
            ReportViewer.entryPlurable,
            entries.length
          )}:`
        )
      );
      missingVaultEntryPrinter(entries);
    };
  }

  private getNoChangesReportHandler(): (
    this: SnapshotReporter
  ) => void | Promise<void> {
    return function (this: SnapshotReporter) {
      CliUx.ux.log(dedent`${green('No resources to change')}.

      The target organization already matches the configuration.`);
      return;
    };
  }

  private getReportErrorsHandler(): (
    this: SnapshotReporter
  ) => void | Promise<void> {
    const reportHandler = this.handleReportErrors.bind(this);
    return function (this: SnapshotReporter) {
      reportHandler(this.resourceInError, this.resourceInErrorCount);
    };
  }

  private printTable() {
    const changedResources = this.reporter.getChangedResources(
      this.operationsToDisplay
    );

    if (changedResources.length === 0) {
      CliUx.ux.log(ReportViewerStyles.header('\nNo changes detected'));
      return;
    }

    CliUx.ux.table(recordable(changedResources), {
      resourceName: {
        header: ReportViewerStyles.header('\nPreviewing resource changes:'),
        get: (resource) => this.createSection(resource),
      },
    });
  }

  private createSection(resource: ReportViewerResourceReportModel) {
    const ResourceSnapshotType = this.prettyPrintResourceName(resource.name);
    const indentation = 3;
    const section = new ReportViewerSection(resource, this.operationsToDisplay);
    let output = `${''.padStart(indentation)}${ResourceSnapshotType}\n`;

    output += section.display(indentation + 1);
    return output;
  }

  private prettyPrintResourceName(resourceName: string): string {
    const capitalized =
      resourceName.charAt(0) + resourceName.slice(1).toLowerCase() + 's';
    return capitalized.replace(/_/g, ' ');
  }

  private handleReportErrors(
    allErrors: Map<Partial<ResourceSnapshotType>, Set<string>>,
    totalCount: number
  ) {
    CliUx.ux.log(ReportViewerStyles.header('Error Report:'));
    CliUx.ux.log(
      ReportViewerStyles.error(
        `   ${totalCount} ${pluralizeIfNeeded(
          ReportViewer.resourcePlurable,
          totalCount
        )} in error `
      )
    );

    for (const [resourceType, errorOfThisResourceType] of allErrors) {
      this.logResourceErrors(resourceType, errorOfThisResourceType);
    }
    this.printNewLine();
  }

  private logResourceErrors(
    resourceSnapshotType: string,
    errorOfThisResourceType: Set<string>
  ) {
    CliUx.ux.log(`\n ${this.prettyPrintResourceName(resourceSnapshotType)}`);
    ReportViewer.printAbridgedMessages(
      Array.from(errorOfThisResourceType),
      ReportViewer.errorPlurable,
      red
    );
  }

  private static printAbridgedMessages(
    messages: string[],
    plurable: Plurable,
    chalker: ChalkFunction
  ) {
    let remainingErrorsToPrint = ReportViewer.maximumNumberOfErrorsToPrint;
    for (let j = 0; j < messages.length && remainingErrorsToPrint > 0; j++) {
      CliUx.ux.log(chalker(`  â€¢ ${messages[j]}`));
      remainingErrorsToPrint--;
    }

    const unprintedMessages =
      messages.length - ReportViewer.maximumNumberOfErrorsToPrint;
    if (unprintedMessages > 0) {
      CliUx.ux.log(
        italic(
          `  (${unprintedMessages} more ${pluralizeIfNeeded(
            plurable,
            unprintedMessages
          )})`
        )
      );
    }
  }

  private printNewLine() {
    CliUx.ux.log('');
  }
}
