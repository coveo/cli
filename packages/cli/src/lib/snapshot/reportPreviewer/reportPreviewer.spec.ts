import {test} from '@oclif/test';

import {ResourceSnapshotsReportType} from '@coveord/platform-client';
import {ReportViewer} from './reportPreviewer';
import dedent from 'ts-dedent';
import {SnapshotReporter} from '../snapshotReporter';
import {
  getErrorReport,
  getReportWithNoProcessedResources,
  getSuccessReport,
} from '../../../__stub__/resourceSnapshotsReportModel';

describe('ReportViewer', () => {
  describe('when the report contains errors', () => {
    let viewer: ReportViewer;
    beforeAll(() => {
      const reporter = new SnapshotReporter(
        getErrorReport('some-id', ResourceSnapshotsReportType.DryRun)
      );
      viewer = new ReportViewer(reporter);
    });

    test
      .stdout()
      .do(() => {
        viewer.display();
      })
      .it(
        'should print a report section with the resources in error',
        (ctx) => {
          expect(ctx.stdout).toContain('Error Report:');
          expect(ctx.stdout).toContain('7 resources in error');
        }
      );

    test
      .stdout()
      .do(() => {
        viewer.display();
      })
      .it('should not print more than 5 errors per resources', (ctx) => {
        expect(ctx.stdout).toContain(
          dedent`
          Fields
            • RESOURCE_ALREADY_EXISTS: Field foo already exists.
            • RESOURCE_ALREADY_EXISTS: Field bar already exists.
            • RESOURCE_ALREADY_EXISTS: Field dsads already exists.
            • RESOURCE_ALREADY_EXISTS: Field fdww already exists.
            • RESOURCE_ALREADY_EXISTS: Field csad already exists.
            (2 more errors)`
        );
      });
  });

  describe('when the report does not contain errors', () => {
    let viewer: ReportViewer;
    beforeAll(() => {
      const reporter = new SnapshotReporter(
        getSuccessReport('some-id', ResourceSnapshotsReportType.DryRun)
      );
      viewer = new ReportViewer(reporter);
    });

    test
      .stdout()
      .do(() => {
        viewer.display();
      })
      .it('should print resource changes', (ctx) => {
        // Remove padding added by cli-ux so we can test the text and not the padding on the line
        const trimedStdout = ctx.stdout
          .split(/$/m)
          .map((s) => s.trimEnd())
          .join('');

        expect(trimedStdout).toContain(dedent`
        Previewing resource changes:
           Extensions
        +   1 to create
        -   2 to delete
        ~   1 to update
           Fields
        ~   1 to update
        `);
      });
  });

  describe('when the report contains no changes', () => {
    let viewer: ReportViewer;
    beforeAll(() => {
      const reporter = new SnapshotReporter(
        getReportWithNoProcessedResources(
          'some-id',
          ResourceSnapshotsReportType.DryRun
        )
      );
      viewer = new ReportViewer(reporter);
    });

    test
      .stdout()
      .do(() => {
        viewer.display();
      })
      .it('should show that no changes were detected', (ctx) => {
        expect(ctx.stdout).toContain('No changes detected');
      });
  });
});
