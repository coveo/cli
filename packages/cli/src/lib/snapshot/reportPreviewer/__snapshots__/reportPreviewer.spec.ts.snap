// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportViewer when the report contains errors should print the same report 1`] = `
" 
Previewing resource changes:
 ──────────────────────────── 
    Extensions                
 +   1 to create              
 -   2 to delete              
                              
    Fields                    
 !   7 in error               
 ~   1 to update              
                              
Error Report:
   7 resources in error 

 Fields
  • RESOURCE_ALREADY_EXISTS: Field foo already exists.
  • RESOURCE_ALREADY_EXISTS: Field bar already exists.
  • RESOURCE_ALREADY_EXISTS: Field dsads already exists.
  • RESOURCE_ALREADY_EXISTS: Field fdww already exists.
  • RESOURCE_ALREADY_EXISTS: Field csad already exists.
  (2 more errors)

"
`;

exports[`ReportViewer when the report contains missing vault entries should print the same report 1`] = `
" 
Previewing resource changes:
 ──────────────────────────── 
    Extensions                
 +   1 to create              
 -   2 to delete              
                              
    Fields                    
 !   7 in error               
 ~   1 to update              
                              
Missing vault entries:
  • VAULT.foo_4VNj5ds5-configuration.userIdentities.UserIdentity.password
  • VAULT.bar_4VNj5ds5-configuration.userIdentities.UserIdentity.password
  • VAULT.dsads_4VNj5ds5-configuration.userIdentities.UserIdentity.password
  • VAULT.fdww_4VNj5ds5-configuration.userIdentities.UserIdentity.password
  • VAULT.csad_4VNj5ds5-configuration.userIdentities.UserIdentity.password
  (2 more entries)
"
`;

exports[`ReportViewer when the report contains no changes should show that no changes were detected 1`] = `
"
No changes detected
No resources to change.

The target organization already matches the configuration.
"
`;

exports[`ReportViewer when the report does not contain errors should print resource changes 1`] = `
" 
Previewing resource changes:
 ──────────────────────────── 
    Extensions                
 +   1 to create              
 -   2 to delete              
 ~   1 to update              
                              
    Fields                    
 ~   1 to update              
                              
"
`;
